nix/all-toplevel-derivations.nix

```diff
@@ -0,0 +1,54 @@
+{ lib
+, pkgSet
+, fn
+, recursionDepth
+, keyFilter
+, ...
+}:
+let
+  go = depth: set':
+    let
+      canSet = builtins.tryEval set';
+    in
+    if canSet.success && builtins.isAttrs canSet.value
+    then
+      let
+        set = canSet.value;
+      in
+      (
+        if (builtins.tryEval (lib.isDerivation set)).value
+        then
+          let
+            meta = builtins.tryEval (fn set);
+          in
+          builtins.deepSeq meta (
+            builtins.trace ("reached leaf: " + toString set)
+              (
+                if meta.success
+                then [ meta.value ]
+                else builtins.trace "package didn't evaluate" [ ]
+              )
+          )
+        else if depth >= recursionDepth
```

Comment:
Hm, is this correct? Shouldn't you always check on the recursion depth (and not just as fallback)? :thinking:

Reply:
But we only recur in the else branch? I’m missing the issue with this.

Reply:
Ah, I see.